### ISSUE TITLE
Inductor rewrites masked encoder attention (matmul+mask+softmax+matmul) into SDPA and silently drops attn_mask

### ISSUE THREAD

**Issue by gkumbhat** [torch-spyre/torch-spyre#4526]
### What happened?

When `torch.compile` (Inductor, `dynamic=False`) can see the full decomposed attention pattern —
`matmul(Q, Kᵀ) → additive mask → softmax → matmul(P, V)` — inside a single compiled graph on the
Spyre backend, Inductor rewrites it into an internal `scaled_dot_product_attention`-equivalent
kernel, and that rewrite **silently drops the additive mask**. There is no error, no exception, and
no `FallbackWarning` — the op runs and returns confident garbage for any padded/masked row.

We hit this in `spyre-inference`'s encoder (bidirectional, no-KV-cache) attention backend. It forced
a real architecture change: the QKᵀ matmul and the mask-add+softmax+PV matmul are now deliberately
compiled as **two separate graphs** (`_packed_qk_matmul` then `_packed_pv` in
[`spyre_encoder_attn.py`](https://github.com/torch-spyre/spyre-inference/blob/main/spyre_inference/v1/attention/backends/spyre_encoder_attn.py#L288-L346))
specifically so Inductor never sees the fusable single-graph pattern. From the code:

```python
def _packed_masked_attention(query, key, value, mask, scale):
    """Scatter-path attention. Compile QK separately from mask + softmax + P·V.

    Compiling ``matmul + mask`` lets Inductor rewrite to ``F.sdpa``, which drops
    ``attn_mask`` on Spyre -- so the mask must stay out of
    the Q·Kᵀ graph. It does live in the P·V graph, which cannot form that
    pattern; see ``_packed_pv``.
    """
```

The evidence for the drop, as measured against `BAAI/bge-reranker-v2-m3` (a masked, non-full-bucket
prompt) when the two ops were compiled together: **cosine similarity ~0.46** against the correct
(unmasked-pattern-avoided) output — i.e. the masked and effectively-unmasked outputs are nearly
uncorrelated, consistent with the mask simply not being applied.

### What did you expect to happen?

Compiling `matmul → mask-add → softmax → matmul` as one graph should produce the same result as
compiling it in pieces (or as calling `F.scaled_dot_product_attention` with an explicit
`attn_mask=`) — within fp16 tolerance. If the fused-rewrite path can't support a live mask, it
should raise or emit a `FallbackWarning`, not silently drop it, per the same principle raised in
#4331.

### How can we reproduce it?

Plain `torch` + `torch_spyre`, no vLLM, no `spyre-inference`:

```python
import torch
import torch.nn.functional as F
import torch_spyre  # noqa: F401

DEV = torch.device("spyre")
B, H, L, D = 1, 12, 512, 64
REAL = 345  # live pad: 345 real rows, 167 pad rows
DTYPE = torch.float16
SCALE = D**-0.5


def cpu_reference(q, k, v):
    """float32 attention over the real rows only -- pad never participates."""
    qs, ks, vs = q[:REAL].float(), k[:REAL].float(), v[:REAL].float()
    scores = torch.einsum("qhd,khd->hqk", qs, ks) * SCALE
    probs = torch.softmax(scores, dim=-1)
    return torch.einsum("hqk,khd->qhd", probs, vs)


def one_graph_masked_attention(q, k, v, mask, scale):
    """QKᵀ, mask-add, softmax, PV -- all inside ONE compiled graph."""
    qq = q.unsqueeze(0).transpose(1, 2)
    kk = k.unsqueeze(0).transpose(1, 2)
    vv = v.unsqueeze(0).transpose(1, 2)
    scores = torch.matmul(qq, kk.transpose(-2, -1)) * scale + mask
    probs = torch.softmax(scores, dim=-1)
    out = torch.matmul(probs, vv)
    t, h, d = q.shape
    return out.transpose(1, 2).reshape(t, h, d)


def two_graph_masked_attention(q, k, v, mask, scale):
    """Same math, QK and mask+softmax+PV compiled as two separate graphs."""
    qq = q.unsqueeze(0).transpose(1, 2)
    kk = k.unsqueeze(0).transpose(1, 2)
    qk = torch.compile(lambda a, b: torch.matmul(a, b.transpose(-2, -1)) * scale, dynamic=False)
    scores = qk(qq, kk)

    def pv(scores_, mask_, v_):
        vv = v_.unsqueeze(0).transpose(1, 2)
        s = scores_ + mask_
        probs = torch.softmax(s, dim=-1)
        return torch.matmul(probs, vv)

    out = torch.compile(pv, dynamic=False)(scores, mask, v)
    t, h, d = q.shape
    return out.transpose(1, 2).reshape(t, h, d)


torch.manual_seed(0)
q = torch.randn(L, H, D, dtype=DTYPE)
k = torch.randn(L, H, D, dtype=DTYPE)
v = torch.randn(L, H, D, dtype=DTYPE)
# Pad rows carry large garbage, so a dropped mask pollutes the real rows visibly.
q[REAL:] = 8.0
k[REAL:] = 8.0
v[REAL:] = -8.0

ref = cpu_reference(q, k, v)  # [REAL, H, D] float32

neg = torch.finfo(DTYPE).min
kv_ok = torch.arange(L) < REAL
row = torch.where(kv_ok, torch.zeros((), dtype=DTYPE), torch.tensor(neg, dtype=DTYPE))
mask = row.unsqueeze(0).expand(L, L).contiguous().unsqueeze(0).unsqueeze(0).to(DEV)

qd, kd, vd = (t.to(DEV) for t in (q, k, v))

for name, fn in (
    ("one graph (QK+mask+softmax+PV fused)", one_graph_masked_attention),
    ("two graphs (QK, then mask+softmax+PV)", two_graph_masked_attention),
):
    compiled = torch.compile(fn, dynamic=False) if name.startswith("one") else fn
    out = compiled(qd, kd, vd, mask, SCALE).to("cpu").float()[:REAL]
    cos = F.cosine_similarity(out.reshape(-1), ref.reshape(-1), dim=0).item()
    print(f"{name:42s} cosine={cos:.4f}  {'MASK DROPPED' if cos < 0.9 else 'mask honoured'}")
```

Expected: both print `cosine≈1.0`. Observed (measured in production against
`BAAI/bge-reranker-v2-m3`'s masked/short-prompt path, same pattern as the "one graph" variant
above): the single-graph path collapses to `cosine≈0.46`; splitting into two compiled graphs
(as `spyre-inference` now does) is the confirmed-correct path.

### Any environmental details we need to know?

- `torch-spyre` git rev pinned by `spyre-inference` at time of filing:
  `c3d949a933b81121a9d1119ccbf73425b6fd3835`
- `torch.compile(dynamic=False)`, Inductor backend, `fp16`
- Encoder (bidirectional, no-KV-cache) attention shapes, e.g. `B=1, H=12, L=512, D=64`

### Anything else we need to know?

- #4331 (closed) — same *symptom class* (SDPA silently wrong on Spyre, no error/warning) but a
  different trigger: non-contiguous q/k/v views, no mask involved.
- #4339 (closed, fixed by #4424) — same *area* (encoder fused SDPA on `torch-spyre`) but a pure
  performance regression on the unmasked path, not a correctness bug.
- #3113 (open) — SDPA mis-lowering, but triggered by non-stick-aligned sequence length, not by
  Inductor's matmul+mask→sdpa rewrite.

None of the above cover this specific rewrite-drops-the-mask behavior, which is why we're filing
separately.

### Relevant log output

```Shell

```

**Comment by dmholtz** [comment]
@gkumbhat I tried to reproduce the bug with the minimal script from above using torch-spyre `c3d949a933b81121a9d1119ccbf73425b6fd3835` and get:

```text
one graph (QK+mask+softmax+PV fused)       cosine=1.0000  mask honoured
two graphs (QK, then mask+softmax+PV)      cosine=1.0000  mask honoured
```

This is what we expect and means no bug, doesn't it? Could you please double check?

**Comment by gkumbhat** [comment]
Hi @dmholtz , thanks for trying to repro. That response means no bug. I also tried it on my end and was not able to repro on latest torch-spyre. So this might have gotten fixed already.

I'll close this issue. Thank you.
